lte/gateway/python/magma/pipelined/qos/qos_tc_impl.py

```diff
@@ -53,42 +51,65 @@ class TrafficClass:
     Creates/Deletes queues in linux. Using Qdiscs for flow based
     rate limiting(traffic shaping) of user traffic.
     """
+    ROOT_QID = 'fffe'
+
     @staticmethod
-    def delete_class(intf: str, qid: int, throw_except=True,
-                     show_error=True) -> None:
+    def delete_class(intf: str, qid: int, is_leaf=True, show_error=True) -> None:
         qid_hex = hex(qid)
-        tc_cmd = "tc class del dev {} classid 1:{}".format(intf, qid_hex)
-        filter_cmd = "tc filter del dev {} protocol ip parent 1: prio 1 \
-                handle {qid} fw flowid 1:{qid}".format(intf, qid=qid_hex)
-        run_cmd((filter_cmd, tc_cmd), throw_except, show_error)
+
+        # delete filter if this is a leaf class
+        if is_leaf:
+            filter_cmd = "tc filter del dev {intf} protocol ip parent 1: prio 1 \
+                    handle {qid} fw flowid 1:{qid}".format(intf=intf, qid=qid_hex)
+            run_cmd([filter_cmd], show_error)
+
+        # delete class
+        tc_cmd = "tc class del dev {intf} classid 1:{qid}".format(intf=intf, 
+            qid=qid_hex)
+        run_cmd([tc_cmd], show_error)
+    
 
     @staticmethod
-    def create_class(intf: str, qid: int, max_bw: int, throw_except=True,
-                     show_error=True) -> None:
+    def create_class(intf: str, qid: int, max_bw: int, parent_qid=None,
+        is_leaf=True, show_error=True) -> None:
         qid_hex = hex(qid)
-        tc_cmd = "tc class add dev {} parent 1:fffe classid 1:{} htb \
-        rate 12000 ceil {}".format(intf, qid_hex, max_bw)
-        qdisc_cmd = "tc qdisc add dev {} parent 1:{} \
-                fq_codel".format(intf, qid_hex)
-        filter_cmd = "tc filter add dev {} protocol ip parent 1: prio 1 \
-                handle {qid} fw flowid 1:{qid}".format(intf, qid=qid_hex)
+        parent_qid_hex = hex(parent_qid) if parent_qid else TrafficClass.ROOT_QID
+        tc_cmd = "tc class add dev {intf} parent 1:{parent_qid} classid 1:{qid} htb \
+            rate 12Kbit ceil {maxbw}".format(intf=intf, parent_qid=parent_qid_hex, 
```

Comment:
let's have the rate parameter also as a function argument. We can use it for GBR where gbr rates can be used for the rate parameter.

Reply:
I added these as default args but it failed due to invalid inputs.. I will do a cleaner check at the top of the method.